File: native/cocos/primitive/Torus.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "primitive/Torus.h"

namespace cc {
IGeometry torus(float radius, float tube, const ccstd::optional<ITorusOptions> &opts) {
    const uint32_t radialSegments = opts.has_value() ? opts->radialSegments : 32;
    const uint32_t tubularSegments = opts.has_value() ? opts->tubularSegments : 32;
    const float arc = opts.has_value() ? opts->arc : math::PI_2;

    ccstd::vector<float> positions;
    ccstd::vector<float> normals;
    ccstd::vector<float> uvs;
    ccstd::vector<uint32_t> indices;
    const Vec3 minPos(-radius - tube, -tube, -radius - tube);
    const Vec3 maxPos(radius + tube, tube, radius + tube);
    const float boundingRadius = radius + tube;

    for (uint32_t j = 0; j <= radialSegments; j++) {
        for (uint32_t i = 0; i <= tubularSegments; i++) {
            const float u = static_cast<float>(i) / static_cast<float>(tubularSegments);
            const float v = static_cast<float>(j) / static_cast<float>(radialSegments);

            const float u1 = u * arc;
            const float v1 = v * math::PI_2;

            // vertex
            const float x = (radius + tube * cos(v1)) * sin(u1);
            const float y = tube * sin(v1);
            const float z = (radius + tube * cos(v1)) * cos(u1);

            // this vector is used to calculate the normal
            const float nx = sin(u1) * cos(v1);
            const float ny = sin(v1);
            const float nz = cos(u1) * cos(v1);

            positions.emplace_back(x);
            positions.emplace_back(y);
            positions.emplace_back(z);
            normals.emplace_back(nx);
            normals.emplace_back(ny);
            normals.emplace_back(nz);

            uvs.emplace_back(u);
            uvs.emplace_back(v);

            if ((i < tubularSegments) && (j < radialSegments)) {
                const uint32_t seg1 = tubularSegments + 1;
                const uint32_t a = seg1 * j + i;
                const uint32_t b = seg1 * (j + 1) + i;
                const uint32_t c = seg1 * (j + 1) + i + 1;
                const uint32_t d = seg1 * j + i + 1;

                indices.emplace_back(a);
                indices.emplace_back(d);
                indices.emplace_back(b);

                indices.emplace_back(d);
                indices.emplace_back(c);
                indices.emplace_back(b);
            }
        }
    }

    IGeometry info;
    info.positions = positions;
    info.normals = normals;
    info.uvs = uvs;
    info.boundingRadius = boundingRadius;
    info.minPos = minPos;
    info.maxPos = maxPos;
    info.indices = indices;
    return info;
}

} // namespace cc

File: cocos/primitive/torus.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Vec3 } from '../core';
import { IGeometryOptions } from './define';

/**
 * @zh
 * 环面参数选项。
 * @en
 * The options about torus
 */
interface ITorusOptions extends IGeometryOptions {
    radialSegments: number;
    tubularSegments: number;
    arc: number;
}

/**
 * @en
 * Generate a torus with radius 0.4, tube 0.1 and centered at origin.
 * @zh
 * 生成一个环面。
 * @param radius @zh 环面半径。@en The radius fo torus
 * @param tube @zh 管形大小。@en The radius of tube
 * @param opts @zh 参数选项。@en The optional creation parameters of the torus
 */
export default function torus (radius = 0.4, tube = 0.1, opts: RecursivePartial<ITorusOptions> = {}): { positions: number[]; normals: number[]; uvs: number[]; indices: number[]; minPos: Vec3; maxPos: Vec3; boundingRadius: number; } {
    const radialSegments = opts.radialSegments || 32;
    const tubularSegments = opts.tubularSegments || 32;
    const arc = opts.arc || 2.0 * Math.PI;

    const positions: number[] = [];
    const normals: number[] = [];
    const uvs: number[] = [];
    const indices: number[] = [];
    const minPos = new Vec3(-radius - tube, -tube, -radius - tube);
    const maxPos = new Vec3(radius + tube, tube, radius + tube);
    const boundingRadius = radius + tube;

    for (let j = 0; j <= radialSegments; j++) {
        for (let i = 0; i <= tubularSegments; i++) {
            const u = i / tubularSegments;
            const v = j / radialSegments;

            const u1 = u * arc;
            const v1 = v * Math.PI * 2;

            // vertex
            const x = (radius + tube * Math.cos(v1)) * Math.sin(u1);
            const y = tube * Math.sin(v1);
            const z = (radius + tube * Math.cos(v1)) * Math.cos(u1);

            // this vector is used to calculate the normal
            const nx = Math.sin(u1) * Math.cos(v1);
            const ny = Math.sin(v1);
            const nz = Math.cos(u1) * Math.cos(v1);

            positions.push(x, y, z);
            normals.push(nx, ny, nz);
            uvs.push(u, v);

            if ((i < tubularSegments) && (j < radialSegments)) {
                const seg1 = tubularSegments + 1;
                const a = seg1 * j + i;
                const b = seg1 * (j + 1) + i;
                const c = seg1 * (j + 1) + i + 1;
                const d = seg1 * j + i + 1;

                indices.push(a, d, b);
                indices.push(d, c, b);
            }
        }
    }

    return {
        positions,
        normals,
        uvs,
        indices,
        minPos,
        maxPos,
        boundingRadius,
    };
}
